Crash when destroying depoyment

#  Version Reports:

### Distro  version of host:

``` openSUSE 42.2 ```

### Terraform Version Report
Terraform v0.9.4

### Libvirt version

```virsh --version```
2.0.0
### terraform-provider-libvirt plugin version (git-hash)

Installed through YAST version: 0.0.0+git20170702.612d0af3-1.2
___
# Description of Issue/Question

### Setup
Deploying CAASP (https://github.com/inercia/caasp-tf).
See the following crashing when I attempt to tear down my setup.

```
Do you really want to destroy?
  Terraform will delete all your managed infrastructure.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

null_resource.local_checkout_of_caasp_image: Refreshing state... (ID: 2556970206911593677)
data.template_file.admin_cloud_init_user_data: Refreshing state...
libvirt_volume.base_img: Refreshing state... (ID: /var/lib/libvirt/images/caasp.qcow2)
libvirt_cloudinit.admin: Refreshing state... (ID: /var/lib/libvirt/images/caasp_beta_admi...o;597b29b2-30a4-f0ff-2992-414dadc16f22)
libvirt_volume.node.1: Refreshing state... (ID: /var/lib/libvirt/images/caasp_beta_node_1.qcow2)
libvirt_volume.admin: Refreshing state... (ID: /var/lib/libvirt/images/caasp_beta_admin.qcow2)
libvirt_volume.node.0: Refreshing state... (ID: /var/lib/libvirt/images/caasp_beta_node_0.qcow2)
libvirt_volume.node.2: Refreshing state... (ID: /var/lib/libvirt/images/caasp_beta_node_2.qcow2)
libvirt_domain.admin: Refreshing state... (ID: 78af5334-35b7-4918-aba9-89acc1b77c19)
libvirt_domain.node.0: Refreshing state... (ID: 64edf276-68f2-412b-94c3-de1c15ae1c82)
libvirt_domain.node.1: Refreshing state... (ID: 55138d3c-c068-4e0c-8ffd-ef692049d7df)
libvirt_domain.node.2: Refreshing state... (ID: ac248a3e-2e23-4f1d-bf1b-48c1d3fad99e)
Error refreshing state: 3 error(s) occurred:
* libvirt_domain.node[2]: libvirt_domain.node.2: unexpected EOF
* libvirt_domain.node[1]: libvirt_domain.node.1: unexpected EOF
* libvirt_domain.node[0]: libvirt_domain.node.0: unexpected EOF
panic: runtime error: invalid memory address or nil pointer dereference
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x58a576]
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: 
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: goroutine 73 [running]:
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/libvirt/libvirt-go.(*StorageVol).Free(0x0, 0x0, 0x1211500)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/libvirt/libvirt-go/storage_volume.go:
135 +0x26
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.resourceLibvirtDomainRead(0xc4203962a0, 0xab0e80, 0xc4201de428, 0x120ee00, 0xc4203e36e0)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/resource_libvirt_domain.go:751 +0xe74
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Refresh(0xc4202b69c0, 0xc42006de0$
, 0xab0e80, 0xc4201de428, 0xc4202b2638, 0xc4202d6701, 0x0)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema/re$
ource.go:314 +0x21d
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Refresh(0xc4202a41c0, 0xc42006ddb$
, 0xc42006de00, 0x7f856067b000, 0x0, 0x0)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema/pr$
vider.go:267 +0x91
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Refresh(0xc42029d620, 0xc4$
03c7740, 0xc4203c7c20, 0x0, 0x0)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/plugin/resource_$
rovider.go:510 +0x4e
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: reflect.Value.call(0xc42001f9e0, 0xc4202a21f0, 0x13, 0xbe7587, 0x4, 0xc420274f20, 0x3, 0x3, 0x0, 0x0, ...)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /usr/lib64/go/1.8/src/reflect/value.go:434 +0x91f
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: reflect.Value.Call(0xc42001f9e0, 0xc4202a21f0, 0x13, 0xc42003ff20, 0x3, 0x3, 0x0, 0x0, 0x0)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /usr/lib64/go/1.8/src/reflect/value.go:302 +0xa4
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: net/rpc.(*service).call(0xc4202c34c0, 0xc4202c3480, 0xc4202cd130, 0xc4202d8400, 0xc42029d720, 0xab2300, 0xc4203c7740, 0x16, 0xab2340, 0xc4203c7c2$
, ...)
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /usr/lib64/go/1.8/src/net/rpc/server.go:387 +0x144
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt: created by net/rpc.(*Server).ServeCodec
2017/08/01 09:49:41 [DEBUG] plugin: terraform-provider-libvirt:         /usr/lib64/go/1.8/src/net/rpc/server.go:481 +0x404
2017/08/01 09:49:41 [ERROR] root: eval: *terraform.EvalRefresh, err: libvirt_domain.node.2: unexpected EOF
2017/08/01 09:49:41 [ERROR] root: eval: *terraform.EvalSequence, err: libvirt_domain.node.2: unexpected EOF
2017/08/01 09:49:41 [TRACE] [walkRefresh] Exiting eval tree: libvirt_domain.node[2]
2017/08/01 09:49:41 [ERROR] root: eval: *terraform.EvalRefresh, err: libvirt_domain.node.1: unexpected EOF
2017/08/01 09:49:41 [ERROR] root: eval: *terraform.EvalSequence, err: libvirt_domain.node.1: unexpected EOF
2017/08/01 09:49:41 [TRACE] [walkRefresh] Exiting eval tree: libvirt_domain.node[1]
2017/08/01 09:49:41 [DEBUG] plugin: /usr/bin/terraform-provider-libvirt: plugin process exited
2017/08/01 09:49:41 [ERROR] root: eval: *terraform.EvalRefresh, err: libvirt_domain.node.0: unexpected EOF
2017/08/01 09:49:41 [ERROR] root: eval: *terraform.EvalSequence, err: libvirt_domain.node.0: unexpected EOF
2017/08/01 09:49:41 [TRACE] [walkRefresh] Exiting eval tree: libvirt_domain.node[0]
2017/08/01 09:49:41 [DEBUG] dag/walk: upstream errored, not walking "provider.libvirt (close)"
2017/08/01 09:49:41 [DEBUG] dag/walk: upstream errored, not walking "output.nodes"
2017/08/01 09:49:41 [DEBUG] dag/walk: upstream errored, not walking "root"
2017/08/01 09:49:41 [WARN] plugin: error closing client during Kill: connection is shut down
2017/08/01 09:49:41 [DEBUG] plugin: waiting for all plugin processes to complete...
2017/08/01 09:49:41 [DEBUG] plugin: terraform: template-provider (internal) 2017/08/01 09:49:41 [DEBUG] plugin: waiting for all plugin processes to complete...
2017/08/01 09:49:41 [DEBUG] plugin: terraform: null-provider (internal) 2017/08/01 09:49:41 [DEBUG] plugin: waiting for all plugin processes to complete...
2017/08/01 09:49:41 [DEBUG] plugin: terraform: local-exec-provisioner (internal) 2017/08/01 09:49:41 [DEBUG] plugin: waiting for all plugin processes to complete...
2017/08/01 09:49:41 [DEBUG] plugin: /usr/bin/terraform: plugin process exited
2017/08/01 09:49:41 [DEBUG] plugin: /usr/bin/terraform: plugin process exited
2017/08/01 09:49:41 [DEBUG] plugin: /usr/bin/terraform: plugin process exited
```

--- a/libvirt/resource_libvirt_domain.go
+++ b/libvirt/resource_libvirt_domain.go
@@ -804,11 +804,11 @@ func resourceLibvirtDomainRead(d *schema.ResourceData, meta interface{}) error {
 
 			virVol, err = virPool.LookupStorageVolByName(diskDef.Source.Volume)
 		}
-		defer virVol.Free()
 
 		if err != nil {
 			return fmt.Errorf("Error retrieving volume for disk: %s", err)
 		}
+		defer virVol.Free()
 
 		virVolKey, err := virVol.GetKey()
 		if err != nil {

